builtin/aws/alb/releaser.go

```diff
@@ -42,192 +46,357 @@ func (r *Releaser) StatusFunc() interface{} {
+		vpc = v
+	} else {
+		for _, s := range r.config.Subnets {
+			subnets = append(subnets, aws.String(s))
+		}
+
+		subnetInfo, err := ec2srv.DescribeSubnets(&ec2.DescribeSubnetsInput{
+			SubnetIds: []*string{subnets[0]},
 		})
-	}
+		if err != nil {
+			return err
+		}
 
-	// If there is a port defined, honor it.
-	if r.config.Port != 0 {
-		port = int64(r.config.Port)
+		vpc = subnetInfo.Subnets[0].VpcId
 	}
+	sGroup, err := utils.CreateSecurityGroup(ctx, sess, fmt.Sprintf("%s-incoming", lbName), vpc, int(port))
+	if err != nil {
+		return err
+	}
+	state.Id = *sGroup
 
-	var (
-		lb               *elbv2.LoadBalancer
-		listener         *elbv2.Listener
-		newListener      bool
-		existingListener string
-	)
+	return nil
+}
 
-	if r.config.ListenerARN != "" {
-		existingListener = r.config.ListenerARN
+func (r *Releaser) resourceSecurityGroupDestroy(
+	ctx context.Context,
+	sess *session.Session,
+	state *Resource_SecurityGroup,
+	log hclog.Logger,
+	sg terminal.StepGroup,
+) error {
+	if state.Id != "" {
+		step := sg.Add("Destroying Security Group...")
+		defer step.Abort()
+		log.Debug("deleting security group", "sg-id", state.Id)
+		ec2Svc := ec2.New(sess)
+		for i := 0; i < 20; i++ {
+			_, err := ec2Svc.DeleteSecurityGroup(&ec2.DeleteSecurityGroupInput{
+				GroupId: &state.Id,
+			})
+			if err == nil {
+				step.Done()
+				return nil
+			}
+			// if we encounter an unrecoverable error, exit now.
+			if aerr, ok := err.(awserr.Error); ok {
+				switch aerr.Code() {
+				case "DependencyViolation":
+					time.Sleep(2 * time.Second)
+					continue
+				case "InvalidGroup.NotFound":
+					log.Debug("security group not found", "sg-id", state.Id)
+					return nil
+				default:
+					return err
+				}
+			}
+			return err
+		}
+		step.Done()
 	}
+	log.Debug("no security group id found, continuing")
 
-	tgs := []*elbv2.TargetGroupTuple{
-		{
-			TargetGroupArn: &target.Arn,
-			Weight:         aws.Int64(100),
-		},
-	}
+	return nil
+}
 
-	if existingListener != "" {
+func (r *Releaser) resourceLoadBalancerCreate(
+	ctx context.Context,
+	log hclog.Logger,
+	sess *session.Session,
+	sg terminal.StepGroup,
+	ui terminal.UI,
+	lbName string,
+	port int64,
+	target *TargetGroup,
+	sgState *Resource_SecurityGroup,
+	state *Resource_LoadBalancer,
+) error {
+
+	elbsrv := elbv2.New(sess)
+	// if a custom listner is being used, we assume the Load Balancer is already
+	// created, and we can skip some steps
+	if r.config.ListenerARN != "" {
+		log.Debug("custom listener arn found, skipping load balancer creation")
 		out, err := elbsrv.DescribeListeners(&elbv2.DescribeListenersInput{
-			ListenerArns: []*string{aws.String(existingListener)},
+			ListenerArns: []*string{aws.String(r.config.ListenerARN)},
 		})
 		if err != nil {
-			return nil, err
+			return err
 		}
 
-		listener = out.Listeners[0]
-	} else {
+		if len(out.Listeners) == 0 {
+			return fmt.Errorf("no listeners found for arn: %s", r.config.ListenerARN)
```

Comment:
Could also do a `codes.NotFound` error here to be more explicit

Reply:
I think there's a few other cases of `fmt.Errorf` that we could replace with `status`/`codes` so I'll just comment on these two 👍🏻

Reply:
Thanks! I updated the 2 specific instances you mention in edf7d4dc, I'll go through and look for others